posts/introduction-to-graphql-with-apollo.md

```diff
@@ -0,0 +1,320 @@
+
+```graphql
+interface Media {
+  id: ID!
+  title: String!
+}
+
+type Book implements Media {
+  id: ID!
+  title: String!
+  totalPages: Int
+  author: Person
+}
+
+type Movie implements Media {
+  id: ID!
+  title: String!
+  runTime: Int
+  director: Person
+}
+```
+
+Alternatively, multiple interfaces can be implemented using the following syntax:
+
+```graphql
+type MyObject implements Interface1 & Interface2 {
+  …
+}
+```
+
+Now imagine we wanted to fetch all the `Media` objects, by retrieving both common and specific fields of the objects implementing this interface.
+
+How does the server know the difference between a `Book` and a `Movie`?
+
+Firstly, the _query_ must be declared in the schema and defined in the associate _resolver_ (such as the **books’** _query_). Next, the special function `__resolveType` provided by apollo-server will be included in the resolver of the `Media` object. This function will allow us to define the type of implementation that will be returned according to the fields available in the object: in the case of `author` field is present, then a `Book` will be returned; otherwise, if the object has a `director` field, a `Film` will be returned.
+
+![interface](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2022/08/graphql.interface.png)
+
+At last, the client-side query will use the `… on Book` or `… on Movie` syntax to retrieve the specific fields to these objects. This notation is called an _inline fragment_.
+
+# Fragments
+
+In addition to _inline fragments_, there are fragments that **allow sharing pieces of query in the client-side**. In the previous example, the `author` and `director` objects have the same _id_, _firstName_ and _lastName_ fields. These fields could be gathered in a `PersonFragment` fragment together:
+
+![fragment](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2022/08/graphql.fragment.png)
+
+# Mutation
+
+_Mutations_ **relate to all data changes: add, edit, delete**. In the same way as _queries_, the _mutation_ must be defined in the schema (in the `Mutation` type) and implemented in the associate function of the _resolver_ (in `Mutation` object). It can also return an object, which can be useful to retrieve the state of the object updated by this _mutation_. However, unlike _queries_, _mutations_ are performed in series, one after the other.
+
+A book's addition looks like this:
+
+![mutation_add](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2022/08/graphql.mutation.add.png)
+
+As in the example, _input_ items can be employed (`BookInput` and `PersonInput`), where each field stands for a parameter, which may be convenient if you want to modify a complex object (instead of modifying fields one by one).
```

Comment:
```suggestion
As in the example, _input_ items can be employed (`BookInput` and `PersonInput`), where each field stands for a parameter, which may be convenient if you want to modify a complex object instead of modifying fields one by one.
```